src/WebJobs.Script.WebHost/Security/KeyManagement/SimpleKubernetesClient.cs

```diff
@@ -0,0 +1,307 @@
+using Microsoft.Azure.WebJobs.Script.IO;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost
+{
+    public class SimpleKubernetesClient : IKubernetesClient
+    {
+        private const string NamespaceFile = "/run/secrets/kubernetes.io/serviceaccount/namespace";
+        private const string TokenFile = "/run/secrets/kubernetes.io/serviceaccount/token";
+        private const string CaFile = "/run/secrets/kubernetes.io/serviceaccount/ca.crt";
+        private const string KubernetesSecretsDir = "/run/secrets/functions-keys";
+        private readonly HttpClient _httpClient;
+        private readonly IEnvironment _environment;
+        private Action _watchCallback;
+        private bool _disposed;
+        private AutoRecoveringFileSystemWatcher _fileWatcher;
+
+        public SimpleKubernetesClient(IEnvironment environment) : this(environment, CreateHttpClient())
+        { }
+
+        // for testing
+        internal SimpleKubernetesClient(IEnvironment environment, HttpClient client)
+        {
+            _httpClient = client;
+            _environment = environment;
+            Task.Run(() => RunWatcher());
+        }
+
+        private string KubernetesObjectName => _environment.GetEnvironmentVariable(EnvironmentSettingNames.AzureWebJobsKubernetesSecretName);
+
+        public bool IsWritable => !string.IsNullOrEmpty(KubernetesObjectName);
+
+        public async Task<IDictionary<string, string>> GetSecrets()
+        {
+            if (string.IsNullOrEmpty(KubernetesObjectName) && FileUtility.DirectoryExists(KubernetesSecretsDir))
+            {
+                return await GetFromFiles(KubernetesSecretsDir);
+            }
+            else if (!string.IsNullOrEmpty(KubernetesObjectName))
+            {
+                return await GetFromApiServer(KubernetesObjectName);
+            }
+            else
+            {
+                throw new InvalidOperationException($"{nameof(KubernetesSecretsRepository)} requires setting {EnvironmentSettingNames.AzureWebJobsKubernetesSecretName} or mounting secrets to {KubernetesSecretsDir}");
+            }
+        }
+
+        public async Task UpdateSecrets(IDictionary<string, string> data)
+        {
+            (var url, var isSecret) = await GetObjectUrl(KubernetesObjectName);
+            await CreateIfDoesntExist(url, isSecret);
+
+            data = isSecret
+                ? data.ToDictionary(k => k.Key, v => Convert.ToBase64String(Encoding.UTF8.GetBytes(v.Value)))
+                : data;
+
+            using (var request = await GetRequest(HttpMethod.Patch, url, new[] { new { op = "replace", path = "/data", value = data } }, "application/json-patch+json"))
+            {
+                var response = await _httpClient.SendAsync(request);
+                response.EnsureSuccessStatusCode();
+            }
+        }
+
+        public void OnSecretChange(Action callback)
+        {
+            _watchCallback = callback;
```

Comment:
This could be an `IList<Action>` if we have multiple subscribers. But I thought I'd keep it simple for now.

Reply:
:+1: